src/main/java/it/pagopa/pm/gateway/utils/VPosDocumentBuilder.java

```diff
@@ -0,0 +1,103 @@
+        addElement(null, tag, value);
+    }
+
+    public void addElement(VposRequestEnum elementRoot, Element element) {
+        Element rootElement = findRoot(elementRoot);
+        rootElement.addContent(element);
+    }
+
+    public void addElement(VposRequestEnum elementRoot, VposRequestEnum tag, Object value) {
+        Element rootElement = findRoot(elementRoot);
+        if (tag != null) {
+            if (tag.getType() != null) {
+                if (tag.isMandatory() && value == null) {
+                    throw new IllegalArgumentException("The request tag: " + tag.getTagName() + " is mandatory");
+                }
+                if (value.getClass().equals(tag.getType())) {
+                    Element element = new Element(tag.getTagName());
+                    String textValue = value.toString();
+                    checkTextValue(value, tag, element, textValue);
+                    rootElement.addContent(element);
+                } else {
+                    throw new IllegalArgumentException("The request tag: " + tag.getTagName()
+                            + " need a value of type: " + tag.getType().getSimpleName());
+                }
+            } else {
+                Element element = new Element(tag.getTagName());
+                rootElement.addContent(element);
+                children.put(tag.getTagName(), element);
+            }
+        } else {
+            throw new IllegalArgumentException("Argument tag cannot be null");
+        }
+    }
+
+    private void checkTextValue(Object value, VposRequestEnum tag, Element element, String textValue) {
+        if (textValue.trim().length() > 0) {
+            if (value instanceof Date) {
+                SimpleDateFormat simpleDateFormat = new SimpleDateFormat(tag.getFormat());
+                textValue = simpleDateFormat.format((Date) value);
+            }
+            element.setText(textValue);
+        }
+    }
+
+    public void addBodyElement(VposRequestEnum tag) {
+        addElement(null, tag, null);
+    }
+
+    public void addBodyElement(VposRequestEnum elementRoot, VposRequestEnum tag) {
+        Element rootElement = children.get(elementRoot.getTagName());
+        if (rootElement == null) {
+            throw new IllegalArgumentException("element root invalid");
+        }
+        Element element = new Element(tag.getTagName());
+        rootElement.addContent(element);
+        children.put(tag.getTagName(), element);
+    }
+
+    public Document build() {
+        children.clear();
+        return request;
+    }
+
+    private Element findRoot(VposRequestEnum elementRoot) {
+        Element rootElement = this.root;
+        if (elementRoot != null) {
+            rootElement = children.get(elementRoot.getTagName());
+            if (rootElement == null) {
+                throw new IllegalArgumentException("element root invalid");
+            }
+        }
+        return rootElement;
+    }
```

Comment:
the proposed check is wrong. the check is done on `elementRoot `and not on `rootElement`.
To avoid confusion i changed the name of the variable `rootElement` to `element`